### DISCUSSION TITLE
Migrations + QueryBuilder/EM caveat: how to deal with it?

### DISCUSSION THREAD

**Discussion by hayksaryan** [mikro-orm/mikro-orm#4903] [Q&A] [answered]
It's cool that we're able to use QueryBuilder in migrations:
```
const em = this.getEntityManager();
const books = await em.createQueryBuilder(Book, "book").select(["book.id"]);
```

But there's a caveat. The QueryBuilder relies on the most recent (current) version of entities. So, you might get type errors once you modify the schema further during the next migrations.

For example, I've encountered an issue where the QueryBuilder tried to select a field from a table that at the time of that specific migration did not exist (because the latest version of the entity has that field). 
But this is just an example. Depending on how you manipulate the result inside that migration or how you modify the entity file further, this might lead to various indefinite amounts of type and runtime errors.

How do you usually approach this?
Should one fallback to the underlying Knex instance or driver which is not typesafe and doesn't return mapped data as an escape hatch? Or even not handle complex schema/data migrations inside MikroORM migration and create a script for that?
If so the docs should be updated with relevant note/warning or discouragement of using QueryBuilder/Entitiy Manager inside Migrations: [relevant section in docs](https://mikro-orm.io/docs/migrations#working-with-entitymanager)

**Comment by B4nan** [comment, marked as answer]
I would discourage doing this in the first place, just do the queries yourself, that's what a migration is supposed to be about. It should be predictable and easily reviewable, not dynamic. Will add a note the docs, I've added the EM getter there because people were doing it already but in a wrong way, without the correct transaction context.

**Reply by hayksaryan** [reply]
Got it, thanks for the additional context!

**Reply by B4nan** [reply]
https://mikro-orm.io/docs/migrations#working-with-entitymanager

**Reply by hayksaryan** [reply]
Amazing!
